DTSW-4570-move-dts-vscode-open-to-using-devcontainer-and

Tuxliri opened the pull request:
Moving `dts vscode open` command to use the configurations from `devcontainers.yaml` and `containers.yaml` files.

Tuxliri committed:
DTSW-3817 added ros.distro to devcontainer config

Tuxliri committed:
Fixed incorrect install instructions

Tuxliri committed:
Implementing the devcontainers layer

Tuxliri committed:
Merge branch 'ente-staging' into DTSW-4570-move-dts-vscode-open-to-using-devcontainer-and

Tuxliri committed:
Fixed parsing of docker version

Tuxliri committed:
Implemented dry run mode, fixed call to devel build

Tuxliri committed:
Implemented container configuration extension

Tuxliri committed:
Fix:compose file service name, container rebooting

Tuxliri committed:
Cleanup of null keys

Tuxliri committed:
Docker arguments not used at runtime

Tuxliri committed:
Implemented workspace folder

Tuxliri committed:
Fixing docker volumes

Tuxliri committed:
Handling parsed network mode and ros master

Tuxliri committed:
Enable selection of devcontainer configuration

Tuxliri committed:
Improved logging of errors

Tuxliri committed:
Added command help

Tuxliri committed:
Added support for specified dockerComposeFile

Tuxliri commented on vscode/open/command.py:
```diff
@@ -98,6 +100,33 @@ def command(shell, args: list):
+
+        container_configuration : ContainerConfiguration = project.containers[devcontainer_configuration.container]
+        
+        # If the '__extend__' key is present, the container configuration is extended from the one specified in the '__extend__' key
+        extend_from : List = container_configuration.__dict__.get('__extends__', None)
+        dtslogger.debug(f"Container configurations to extend from: {extend_from}")
+
+        if extend_from is not None:
+            for container_name in extend_from:
+                if container_name not in project.containers:
+                    dtslogger.error(f"The container configuration '{container_name}' you want to extend from does not exist in 'containers.yaml'")
+                    return
+
+                dtslogger.debug(f"Extending container configuration from '{container_name}'")
+                container_configuration.__dict__.update(project.containers[container_name].__dict__)
+            container_configuration.__dict__.pop('__extends__', None)
+
+        container_configuration.__dict__.pop('__plain__', None)
+
+        dtslogger.debug(f"Loaded container configurations: {project.containers.__dict__.items()}")
+        dtslogger.debug(f"Using the '{devcontainer_configuration.container}' container configuration: {container_configuration.__dict__.items()}")
```
The `ContainerConfiguration` class inherits from `dict`, however being a `dataclass` it does really work as a dictionary. Thus here we have to use the `__dict__` attribute of the class, which is not the most elegant. This works but maybe we should revisit it.

afdaniele replied:
Definitely needs to be revisited, we can probably add an `update()` function to the dataclass class that internally does `__dict__.update()`. Also, this logic should be moved to the DTProject class. We want to do `project.get_devcontainer(name: str)` and expect an object of type `DevContainerConfiguration` that is already extended.

Tuxliri commented on vscode/open/command.py:
```diff
@@ -106,18 +135,19 @@ def command(shell, args: list):
-            # Add VEHICLE_NAME as an environment variable of the container
-            DEVCONTAINER_TEMPLATE["containerEnv"]["VEHICLE_NAME"] = parsed.ros
-        
+            if isinstance(container_configuration.environment, dict):
+                # It's a dictionary
+                container_configuration.environment['VEHICLE_NAME'] = parsed.ros
+            elif isinstance(container_configuration.environment, list):
+                # It's a list
+                container_configuration.environment = [env for env in container_configuration.environment if not env.startswith('VEHICLE_NAME=')]
+                container_configuration.environment.append(f'VEHICLE_NAME={parsed.ros}')
+
         # add default mount points
```
Some properties (such as the `environment` one) in the `docker-compose.yml` file can be specified as either a dictionary of `{'key': 'value',}` pairs or as a list `['key=value',]`. This can be problematic to handle in particular when extending a container configuration with the  `__extend__` attribute in `containers.yaml`.

afdaniele replied:
Let's force it to be a dictionary, we don't support the list version.

Tuxliri replied:
ok. Here we set the `VEHICLE_NAME` to be the same as the ros master hostname and `ROS_MASTER_URI` should be set by the entrypoint script from `dt-commons`, however I suspect that this is not getting executed or it is overwritten as those variables do not change in the docker container.

Tuxliri replied:
Implemented here and in the `ContainerConfiguration` class.

afdaniele commented on vscode/open/command.py:
```diff
@@ -171,74 +202,56 @@ def command(shell, args: list):
 
-        # escape spaces in arguments
-        parsed.docker_args = [a.replace(" ", "\\ ") for a in parsed.docker_args]
+        compose_file["services"]={devcontainer_configuration.container : container_configuration.__dict__}
```
no need to use `devcontainer_configuration.container` here. We can call the service always `devcontainer` regardless of the actual configuration, it is just a name. Also, better for when you do `docker ps`, you will see `devcontainer` in the name of the container.

Tuxliri replied:
Ok. We can change this in the future if we want to be able to specify a specific `service` to use as our devcontainer.

afdaniele commented on vscode/open/command.py:
```diff
@@ -98,6 +100,33 @@ def command(shell, args: list):
 
         # get info about project
         project = DTProject(parsed.workdir)
+        
+        try:
+            devcontainer_configuration : DevContainerConfiguration = project.devcontainers[parsed.devcontainer_config]
+        except KeyError as e:
+            dtslogger.error(f"Could not find the {e} configuration in 'devcontainers.yaml'")
+            return
```
If a devcontainer is not defined we don't give an error, we simply run VSCode without devcontainer. We want this to be backward compatible with the MOOC and others LXs that do not have and do not need devcontainers.

Tuxliri replied:
Do you mean simply opening vscode in the current folder or launching the lx with `dts code workbench`?

Tuxliri replied:
`dts code workbench` uses internally the vscode open command, we want to retain the ability to use it to open an instance of vscode.

Also modify the `--devcontainer` command to behave as the `-M` flag in `dts devel run`: if we only pass the flag `--devcontainer` it uses the `default` one and if we pass also a value `--devcontainer=mydevcontainer` we want to open `mydevcontainer`

Tuxliri committed:
Renamed devcontainer service to `dt-devcontainer`

Tuxliri committed:
Handling missing attribute

Tuxliri committed:
Retrieving devcontainer configuration from DTProject

Tuxliri committed:
Graceful handling of missing attributes

Tuxliri committed:
Changing behavior of `--devcontainer` flag

The flag now accepts both str arguments or no arguments. If no arguments are given the default devcontainer configuration is used. If the flag is not given the local folder is opened in an instance of vscode.
